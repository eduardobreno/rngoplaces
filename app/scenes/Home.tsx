import React, { useState, useEffect, useRef } from "react";
import {
  View,
  Alert,
  StyleSheet,
  Dimensions,
  Text,
  Image,
  Platform
} from "react-native";
import { Toast, Button } from "native-base";

import Carousel from "react-native-snap-carousel";
import Geolocation from "@react-native-community/geolocation";
import MapView, { Marker, LatLng, AnimatedRegion } from "react-native-maps";
import axios from "axios";

import I18n from "app/helpers/i18n";
import { calculateBetween } from "app/helpers/distance";
import { goToCoordinate } from "app/helpers/maps";
import { askDefaultPermission } from "app/services/api/permissionAPI.ts";
import { CardCarousel } from "app/components/CardCarousel";
import keyMaps from "app/resources/keyMaps";

const { width: viewportWidth, height: viewportHeight } = Dimensions.get(
  "window"
);

interface IMaker {
  id: string;
  coordinate: LatLng | AnimatedRegion;
  title: string;
  description: string;
}
let isBack = false;
let isNotDragging = true;

export const getPlacesNearby = async (lat: number, lng: number) => {
  let markers = [];
  try {
    const result = await axios.get(
      "https://maps.googleapis.com/maps/api/place/nearbysearch/json",
      {
        params: {
          location: `${lat},${lng}`,
          radius: "1000",
          key: keyMaps
        }
      }
    );
    const places = result.data.results;
    markers = places.map((item: any) => {
      const distance = calculateBetween(
        lat,
        lng,
        item.geometry.location.lat,
        item.geometry.location.lng
      );
      return {
        id: item.id,
        coordinate: {
          latitude: item.geometry.location.lat,
          longitude: item.geometry.location.lng
        },
        photos: item.photos,
        title: item.name,
        description: item.vicinity,
        distance: distance
      };
    });
  } catch (e) {
    Alert.alert(I18n.t("error"), I18n.t("failToFindPlacesNearby"));
  }
  return markers;
};

export const getPosition = async (setCurrPosition: any, setMarkers: any) =>
  Geolocation.getCurrentPosition(
    async position => {
      const { coords } = position;
      setCurrPosition({
        latitude: coords.latitude,
        longitude: coords.longitude,
        latitudeDelta: 0.02,
        longitudeDelta: 0.02
      });
      if (!isBack) {
        const result = await getPlacesNearby(coords.latitude, coords.longitude);
        result.sort((a: any, b: any) => a.distance - b.distance);
        setMarkers(result);
      }
    },
    error => {
      Alert.alert(I18n.t("error"), I18n.t("cantFindYourPosition"));
    },
    { enableHighAccuracy: true, timeout: 15000, maximumAge: 1000 }
  );

export const MarkerList = ({ markers }: any) =>
  markers.map((marker: IMaker) => (
    <Marker
      key={marker.id}
      coordinate={marker.coordinate}
      title={marker.title}
      description={marker.description}
    />
  ));

export const searchByRegion = async (coords: any, setMarkers: any) => {
  const result = await getPlacesNearby(coords.latitude, coords.longitude);
  result.sort((a: any, b: any) => a.distance - b.distance);
  setMarkers(result);
};

const Home = () => {
  const [currPosition, setCurrPosition] = useState();
  const [mapPosition, setMapPosition] = useState();
  const [searchInNewPlace, setSearchInNewPlace] = useState(false);
  const mapViewRef = useRef(null);
  const carouselRef = useRef(null);
  const [markers, setMarkers] = useState<Array<IMaker> | []>([]);

  useEffect(() => {
    askDefaultPermission()
      .then(e => {
        Toast.show({
          text: I18n.t("wait"),
          duration: 2000
        });
        getPosition(setCurrPosition, setMarkers);
      })
      .catch(e => {
        // console.log(e);
      });
  }, []);

  return (
    <View
      style={{
        flex: 1,
        flexDirection: "row",
        backgroundColor: "pink",
        alignItems: "flex-end",
        paddingBottom: 15
      }}>
      {searchInNewPlace && (
        <>
          <Button
            full
            info
            style={style.btnFindMore}
            onPress={() => {
              setSearchInNewPlace(false);
              searchByRegion(mapPosition, setMarkers);
            }}>
            <Text>{I18n.t("findHere")}</Text>
          </Button>
          <Button
            full
            info
            style={style.btnResetPosition}
            onPress={() => {
              isBack = true;
              setSearchInNewPlace(true);
              getPosition(setMapPosition, setMarkers);
            }}>
            <Image
              style={{ width: 20, height: 20 }}
              source={require("app/assets/images/reset-position.png")}
            />
          </Button>
        </>
      )}
      <MapView
        ref={e => {
          //@ts-ignore
          mapViewRef.current = e;
        }}
        initialRegion={currPosition}
        region={mapPosition}
        onPanDrag={e => {
          isNotDragging = false;
        }}
        onRegionChangeComplete={e => {
          if (!isNotDragging) {
            setSearchInNewPlace(true);
            setMarkers([]);
            setMapPosition(e);
            // setCurrPosition(e);
          }
        }}
        style={style.maps}>
        {currPosition && (
          <Marker
            key="user"
            coordinate={currPosition}
            image={require("app/assets/images/location.png")}
          />
        )}
        <MarkerList markers={markers} />
      </MapView>
      {!searchInNewPlace && (
        <Carousel
          ref={(e: any) => {
            //@ts-ignore
            carouselRef.current = e;
          }}
          layout="default"
          data={markers}
          renderItem={item => CardCarousel(item, carouselRef, mapViewRef)}
          sliderWidth={viewportWidth}
          itemWidth={200}
          itemHeight={200}
          loop={false}
          onSnapToItem={item => {
            isNotDragging = true;
            //@ts-ignore
            mapViewRef.current.animateCamera(
              goToCoordinate(markers[item].coordinate),
              { duration: 500 }
            );
          }}
        />
      )}
    </View>
  );
};

const style = StyleSheet.create({
  btnFindMore: {
    position: "absolute",
    left: "27%",
    top: "10%",
    paddingLeft: 10,
    paddingRight: 10
  },
  btnResetPosition: {
    position: "absolute",
    right: "5%",
    bottom: "10%",
    paddingLeft: 10,
    paddingRight: 10
  },
  maps: {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: -1
  }
});

export default Home;
