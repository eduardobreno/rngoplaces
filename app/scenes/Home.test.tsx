import React from "react";
import { mount, shallow } from "enzyme";
//@ts-ignore
import MockAsyncStorage from "mock-async-storage";
import axios from "axios";

import Home, { MarkerList, getPosition, getPlacesNearby } from "./Home";
jest.mock("View", () => "view");
jest.mock("axios");
jest.mock("native-base", () => ({
  Toast: { show: jest.fn() }
}));

jest.mock("react-native-permissions", () => ({
  checkMultiple: jest.fn().mockResolvedValue({ location: "authorized" })
}));

jest.mock("react-native-localize", () => ({
  getLocales: jest.fn()
}));

jest.mock("@react-native-community/geolocation", () => ({
  getCurrentPosition: jest.fn((successPosition, errorPosition) => {
    successPosition(() => {});
    errorPosition({ error: { code: {} } });
  })
}));

jest.mock("react-native-maps", () => {
  const React = require.requireActual("react");

  class MockMarker extends React.Component {
    render() {
      return React.createElement("Marker", this.props, this.props.children);
    }
  }

  class MockMapView extends React.Component {
    render() {
      return React.createElement("map-view", this.props, this.props.children);
    }
  }

  MockMapView.Marker = MockMarker;
  return MockMapView;
});

const mockImpl = new MockAsyncStorage();
jest.mock("@react-native-community/async-storage", () => mockImpl);

describe("Home", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = mount(<Home />);
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("getPosition works", async () => {
    const setCurrPosition = jest.fn();
    const setMarkers = jest.fn();

    await getPosition(setCurrPosition, setMarkers);

    expect(wrapper).toMatchSnapshot();
  });

  it("getPlacesNearby works", async () => {
    //@ts-ignore
    axios.get.mockResolvedValue({
      data: {
        results: [
          {
            geometry: {
              location: { lat: -23.6133058, lng: -46.64603220000001 }
            },
            id: "33ad02e48ec3c4bc49c50d3d8d7fe821de2b3fb7",
            name: "Comodita Homestay",
            vicinity: "Avenida Odila, 935, São Paulo"
          },
          {
            geometry: {
              location: { lat: -23.6075828, lng: -46.6424514 }
            },
            id: "54ca3a39b9662241fdf097de945c820bcf5afa41",
            name: "Metalcan Fotolitos Especiais Ltda.",
            vicinity: "Rua das Orquídeas, 259 - Mirandópolis, São Paulo"
          },
          {
            geometry: {
              location: { lat: -23.6114467, lng: -46.6370655 }
            },
            id: "0ed3cc9164deabed5fa710412224c2e6c4beaee3",
            name: "Magazine Luiza",
            vicinity: "Avenida Bosque da Saúde, 138 - Vila da Saúde, São Paulo"
          }
        ]
      }
    });

    const markers = await getPlacesNearby(0, 0);
    expect(markers).toHaveLength(3);
  });

  it("getPlacesNearby fails ", async () => {
    //@ts-ignore
    axios.get.mockRejectedValue();
    const markers = await getPlacesNearby(0, 0);
    expect(markers).toHaveLength(0);
  });

  it("renders Marker", () => {
    const setCurrPosition = jest.fn();
    const setMarkers = jest.fn();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders MarkersList", () => {
    const json = [
      {
        id: "33ad02e48ec3c4bc49c50d3d8d7fe821de2b3fb7",
        coordinate: { latitude: -23.6133058, longitude: -46.64603220000001 },
        title: "Comodita Homestay",
        description: "Avenida Odila, 935, São Paulo"
      },
      {
        id: "54ca3a39b9662241fdf097de945c820bcf5afa41",
        coordinate: { latitude: -23.6075828, longitude: -46.6424514 },
        title: "Metalcan Fotolitos Especiais Ltda.",
        description: "Rua das Orquídeas, 259 - Mirandópolis, São Paulo"
      },
      {
        id: "0ed3cc9164deabed5fa710412224c2e6c4beaee3",
        coordinate: { latitude: -23.6114467, longitude: -46.6370655 },
        title: "Magazine Luiza",
        description: "Avenida Bosque da Saúde, 138 - Vila da Saúde, São Paulo"
      }
    ];
    const markerList = shallow(<MarkerList markers={json} />);
    //     markerList.findWhere(
    //       n =>
    //         n.name() === "Marker" &&
    //         n.prop("key") === "0ed3cc9164deabed5fa710412224c2e6c4beaee3"
    //     )

    expect(markerList).toMatchSnapshot();
  });
});
