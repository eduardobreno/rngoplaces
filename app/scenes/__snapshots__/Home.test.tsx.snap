// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home getPosition works 1`] = `ReactWrapper {}`;

exports[`Home renders Marker 1`] = `ReactWrapper {}`;

exports[`Home renders MarkersList 1`] = `ShallowWrapper {}`;

exports[`Home renders correctly 1`] = `ReactWrapper {}`;

exports[`Home searchByRegion works 1`] = `ReactWrapper {}`;
